core/src/main/scala/kafka/coordinator/transaction/TransactionMarkerChannel.scala

```diff
@@ -0,0 +1,93 @@
+import org.apache.kafka.common.utils.Utils
+import org.apache.kafka.common.{Node, TopicPartition}
+
+import scala.collection.{concurrent, immutable, mutable}
+import collection.JavaConverters._
+
+class TransactionMarkerChannel(interBrokerListenerName: ListenerName, metadataCache: MetadataCache) extends Logging {
+
+  private[transaction] val brokerStateMap: concurrent.Map[Int, DestinationBrokerAndQueuedMarkers] = concurrent.TrieMap.empty[Int, DestinationBrokerAndQueuedMarkers]
+  private val pendingTxnMap: concurrent.Map[Long, TransactionMetadata] = concurrent.TrieMap.empty[Long, TransactionMetadata]
+
+  def addNewBroker(broker: Node) {
+    val markersQueue = new LinkedBlockingQueue[CoordinatorEpochAndMarkers]()
+    brokerStateMap.put(broker.id, DestinationBrokerAndQueuedMarkers(broker, markersQueue))
+    trace(s"Added destination broker ${broker.id}: $broker")
+  }
+
+  private[transaction] def addRequestForBroker(brokerId: Int, txnMarkerRequest: CoordinatorEpochAndMarkers) {
+    val markersQueue = brokerStateMap(brokerId).markersQueue
+    markersQueue.add(txnMarkerRequest)
+    trace(s"Added markers $txnMarkerRequest for broker $brokerId")
+  }
+
+  def addRequestToSend(pid: Long, epoch: Short, result: TransactionResult, coordinatorEpoch: Int, topicPartitions: immutable.Set[TopicPartition]): Unit = {
+    val partitionsByDestination: immutable.Map[Int, immutable.Set[TopicPartition]] = topicPartitions.groupBy { topicPartition: TopicPartition =>
+      val leaderForPartition = metadataCache.getPartitionInfo(topicPartition.topic, topicPartition.partition)
+      leaderForPartition match {
+        case Some(partitionInfo) =>
+          val brokerId = partitionInfo.leaderIsrAndControllerEpoch.leaderAndIsr.leader
+          if (!brokerStateMap.contains(brokerId)) {
+            val broker = metadataCache.getAliveEndpoint(brokerId, interBrokerListenerName).get
+            addNewBroker(broker)
+          }
+          brokerId
+        case None =>
+          // TODO: there is a rare case that the producer gets the partition info from another broker who has the newer information of the
+          // partition, while the TC itself has not received the propagated metadata update; do we need to block when this happens?
+          throw new IllegalStateException("Cannot find the leader broker for the txn marker to write for topic partition " + topicPartition)
+      }
+    }
+
+    for ((brokerId: Int, topicPartitions: immutable.Set[TopicPartition]) <- partitionsByDestination) {
+      val txnMarker = new TxnMarkerEntry(pid, epoch, result, topicPartitions.toList.asJava)
+      addRequestForBroker(brokerId, CoordinatorEpochAndMarkers(coordinatorEpoch, Utils.mkList(txnMarker)))
+    }
+  }
+
+  def maybeAddPendingRequest(metadata: TransactionMetadata): Unit = {
+    val existingMetadataToWrite = pendingTxnMap.putIfAbsent(metadata.pid, metadata)
+    if (existingMetadataToWrite.isDefined)
```

Comment:
Yes, we should fail the subsequent EndTxnRequest. But it seems that we want the client to backoff a bit and keep retrying until success. If we are not sending back a retriable error, the client will think the EndTxnRequest actually permanently failed, which is not the case here.